### PR TITLE
Upgrade to Java Driver version 4.3.0

### PR DESCRIPTION
**Pull request by chbatey**
* [x] Set idempotency on statements
* [x] Migration guide
* [x] Set profile on all executions

We could add some docs for how to override consistency/retry etc but it would just be replicating the java driver docs. If we decide we want that then i'll follow up in a new PR

Refs https://github.com/akka/akka-persistence-cassandra/issues/491

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -32,8 +32,9 @@ lazy val core = (project in file("core"))
     OsgiKeys.importPackage := Seq(akkaImport(), optionalImport("org.apache.cassandra.*"), "*"),
     OsgiKeys.privatePackage := Nil,
     testOptions in Test ++= Seq(
-        Tests.Argument(TestFrameworks.ScalaTest, "-o"),
-        Tests.Argument(TestFrameworks.ScalaTest, "-h", "target/test-reports")))
+        Tests.Argument(TestFrameworks.ScalaTest, "-o")
```
Comment (chbatey) [inline build.sbt]: FIXME, undo

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala [hunk 2]
```diff
@@ -45,45 +44,46 @@ import scala.concurrent.ExecutionContextExecutor
...
           }
         }
 
+      private def isExhausted(rs: AsyncResultSet): Boolean = {
+        rs.remaining() == 0 && !rs.hasMorePages
+      }
+
       private def query(): Unit = {
         def doQuery(): Unit = {
+          // FIXME, what has happened to fetch size per statement?
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: It is in the profile, I think that is granular enough

File: core/src/main/resources/reference.conf [hunk 3]
```diff
@@ -12,37 +12,37 @@
...
+  session-provider = "akka.cassandra.session.DefaultSessionProvider"
 
-  # The identifier that will be passed as parameter to the
-  # ConfigSessionProvider.lookupContactPoints method. 
-  cluster-id = ""
+  # profile to use, can set separate ones for reads and writes
+  # See https://docs.datastax.com/en/developer/java-driver/4.3/manual/core/configuration/
+  # for overriding any settings
+  read-profile = "cassandra-journal"
+  write-profile = "cassandra-journal"
```
Comment (patriknw) [inline core/src/main/resources/reference.conf]: I guess this PR doesn't intend to solve entire https://github.com/akka/akka-persistence-cassandra/issues/81 ?
We still have cassandra-journal, cassandra-snapshot-store, cassandra-query-journal.

Good to do that as a separate change.

File: docs/src/main/paradox/migrations.md [hunk 4]
```diff
@@ -1,8 +1,26 @@
...
-### Migrations to 1.0
+### Migrating from 0.80+ to 1.0
 
-* After being deprecated, the DateTieredCompactionStrategy has been removed.
+#### Driver upgrade with significant configuration changes
+
+Version 4.x of the cassandra driver comes with a new way to configure it via [typesafe configuration](https://github.com/lightbend/config)
+which matches how Akka manages configuration.
+
+All driver related configuration related to driver connection, query consistency, query retries etc has been removed from this
```
Comment (patriknw) [inline docs/src/main/paradox/migrations.md]: related ... related

File: docs/src/main/paradox/migrations.md [hunk 5]
```diff
@@ -1,8 +1,26 @@
...
+project's `reference.conf` and now each part of the plugin (journal, snapshot and query) specify a read and write 
+[execution profile](https://docs.datastax.com/en/developer/java-driver/4.3/manual/core/configuration/#execution-profiles) that gives
+fine grained control over consistencies and retires for each are. By default all read/write profiles are the same and under
+`datastax-java-driver.profile.cassandra-journal`. The only value in the profile provided by the plugin is setting the `basic.request.consistency`
+to `QUORUM`.
+
+The new driver supports reconnection during initialization which was previously built into the plugin. It is recommended to turn this on with:
+`datastax-java-driver.advanced.reconnect-on-init = true`
+It can't be turned on by the plugin as it is in the driver's reference.conf and is not overridable in a profile.
```
Comment (patriknw) [inline docs/src/main/paradox/migrations.md]: that's an important note that we should also have in the main docs

File: core/src/main/resources/reference.conf (hunk 3 above)
Comment (chbatey) [reply core/src/main/resources/reference.conf]: Yes i'll leave until after, should be easy via an extension

**Comment by chbatey** [comment]
The tests take an age to run now due to https://github.com/akka/akka-persistence-cassandra/issues/597 i've tried to reduce this in the last commit

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraEventUpdate.scala [hunk 6]
```diff
@@ -22,14 +22,10 @@ private[akka] trait CassandraEventUpdate extends CassandraStatements {
   private[akka] implicit val ec: ExecutionContext
   private[akka] val log: LoggingAdapter
 
-  def psUpdateMessage: Future[PreparedStatement] =
-    session.prepare(updateMessagePayloadAndTags).map(_.setIdempotent(true))
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraEventUpdate.scala]: all statements are idempotent so it is just in the default profile now

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraEventUpdate.scala [hunk 7]
```diff
@@ -97,38 +94,33 @@ private[akka] trait CassandraEventUpdate extends CassandraStatements {
...
-      // event update
-      bind.setBytes("event", event.serialized)
-      bind.setString("ser_manifest", event.serManifest)
-      bind.setInt("ser_id", event.serId)
-      bind.setString("event_manifest", event.eventAdapterManifest)
-
-      session.executeWrite(bind)
+      val bound = ps
+        .bind()
+        .setString("tag_name", tag)
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraEventUpdate.scala]: lots of these styles of changes as the builders have become immutable

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalConfig.scala [hunk 8]
```diff
@@ -50,9 +50,12 @@ case class TableSettings(
 class CassandraJournalConfig(system: ActorSystem, config: Config)
     extends CassandraPluginConfig(system, config)
     with NoSerializationVerificationNeeded {
+
+  val writeProfile: String = config.getString("write-profile")
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournalConfig.scala]: it is now via a profile that you can change consistency, retries etc

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraStatements.scala [hunk 9]
```diff
@@ -305,29 +307,42 @@ trait CassandraStatements {
...
+          _ <- session.executeAsync(createTable).toScala
+          _ <- session.executeAsync(createMetadataTable).toScala
           done <- tagStatements
         } yield done
       } else keyspace
+
+      import scala.concurrent.duration._
+      import akka.util.PrettyDuration._
+      done.onComplete { result =>
+        session.setSchemaMetadataEnabled(null)
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraStatements.scala]: TODO: remove this, just working out why the build is so slow now

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala [hunk 10]
```diff
@@ -45,45 +45,45 @@ import scala.concurrent.ExecutionContextExecutor
...
+          rs.currentPage().forEach { row =>
+            val s = row.getString("persistence_id")
             if (!knownPersistenceIds.contains(s)) {
               buffer = buffer.enqueue(s)
               knownPersistenceIds += s
             }
           }
           flush()
-          if (refreshInterval.isEmpty && buffer.isEmpty && rs.isExhausted) {
+          if (refreshInterval.isEmpty && buffer.isEmpty && rs.hasMorePages) {
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: This is a good example to take a look at to see how the semantics of the async driver have changed.

You can't ask for more results until you've finished the existing page and there are no blocking calls on at `AsyncResultSet`

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala [hunk 11]
```diff
@@ -69,39 +70,33 @@ import com.github.ghik.silencer.silent
...
+        selectEventsByPersistenceIdQuery
+          .bind(persistenceId, partitionNr: JLong, progress: JLong, toSeqNr: JLong)
+          .setExecutionProfileName(profile)
       executeStatement(boundStatement)
     }
 
     def selectSingleRow(persistenceId: String, pnr: Long)(implicit ec: ExecutionContext): Future[Option[Row]] = {
-      val boundStatement = selectSingleRowQuery.bind(persistenceId, pnr: JLong)
-      session.executeAsync(boundStatement).asScala.map(rs => Option(rs.one()))
+      val boundStatement = selectSingleRowQuery.bind(persistenceId, pnr: JLong).setExecutionProfileName(profile)
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: It is now important to set the right execution profile

We could have had a `CassandraSession` always use the same one but that would conflict with wanting to share the same CassandraSession for all parts of the plugin

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala [hunk 12]
```diff
@@ -15,20 +15,21 @@ import akka.persistence.PersistentRepr
 import akka.persistence.cassandra.journal.CassandraJournal.{ EventDeserializer, Serialized }
 import akka.serialization.Serialization
 import akka.stream.{ Attributes, Outlet, SourceShape }
-import akka.cassandra.session._
 import akka.stream.stage._
-import com.datastax.driver.core._
-import com.datastax.driver.core.policies.RetryPolicy
-import com.datastax.driver.core.utils.Bytes
+import com.datastax.oss.driver.api.core.cql._
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: These changes are not mechanical, review this file

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala [hunk 13]
```diff
@@ -13,24 +13,27 @@ import akka.persistence.cassandra.journal.{ BucketSize, TimeBucket }
...
 import akka.stream.stage.{ GraphStage, _ }
 import akka.stream.{ ActorMaterializer, Attributes, Outlet, SourceShape }
-import akka.cassandra.session._
 import akka.util.PrettyDuration._
 
 import scala.annotation.tailrec
 import scala.concurrent.duration.Duration
 import scala.concurrent.duration._
-import scala.concurrent.{ ExecutionContext, ExecutionContextExecutor, Future }
+import scala.concurrent.{ ExecutionContextExecutor, Future }
```
Comment (chbatey) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala]: Same for this one, review for new use of the async API

**Comment by chbatey** [comment]
Worth reading https://docs.datastax.com/en/developer/java-driver/4.3/upgrade_guide/
and https://docs.datastax.com/en/developer/java-driver/4.3/manual/core/configuration/ before reviewing

File: akka-26-tests/src/test/scala/akka/persistence/cassandra/journal/CassandraLoadTypedSpec.scala [hunk 14]
```diff
@@ -117,13 +106,10 @@ object CassandraLoadTypedSpec {
 
 }
 
-class CassandraLoadTypedSpec extends CassandraSpec(CassandraLoadTypedSpec.config) with WordSpecLike with Matchers {
+class CassandraLoadTypedSpec extends CassandraSpec() with WordSpecLike with Matchers {
```
Comment (helena) [inline akka-26-tests/src/test/scala/akka/persistence/cassandra/journal/CassandraLoadTypedSpec.scala]:
```suggestion
class CassandraLoadTypedSpec extends CassandraSpec with WordSpecLike with Matchers {
```

File: core/src/main/scala/akka/persistence/cassandra/EventsByTagMigration.scala [hunk 15]
```diff
@@ -24,8 +23,9 @@ import akka.stream.{ ActorMaterializer, OverflowStrategy }
 import akka.stream.scaladsl.{ Sink, Source }
 import akka.util.Timeout
 import akka.{ Done, NotUsed }
-import com.datastax.driver.core.Row
-import com.datastax.driver.core.utils.Bytes
+import com.datastax.oss.driver.api.core.cql.Row
+import com.datastax.oss.protocol.internal.util.Bytes
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/EventsByTagMigration.scala]: interesting (the new `oss` package and move)

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraEventUpdate.scala (hunk 6 above)
Comment (helena) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraEventUpdate.scala]: Nice to remove that boilerplate

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala [hunk 16]
```diff
@@ -129,11 +129,11 @@ import scala.concurrent.ExecutionContextExecutor
...
 
             case Some(rs) =>
-              if (refreshInterval.isEmpty && rs.isExhausted) {
+              if (refreshInterval.isEmpty && isExhausted(rs)) {
                 complete(out)
               } else {
-                if (!queryInProgress && !rs.isFullyFetched) {
-                  rs.fetchMoreResults().asScala.foreach(queryCallback.invoke)
+                if (!queryInProgress && rs.remaining() == 0 && rs.hasMorePages) {
+                  rs.fetchNextPage().thenAccept(queryCallback.invoke)
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: The move to more/next page from fetch/exhausted.

File: docs/src/main/paradox/migrations.md [hunk 17]
```diff
@@ -1,8 +1,26 @@
...
+
+#### Removals
+
+Using DateTieredCompactionStrategy with automatic schema creation
```
Comment (helena) [inline docs/src/main/paradox/migrations.md]: Why is this here when it was already merged here https://github.com/akka/akka-persistence-cassandra/pull/590?

File: core/src/test/scala/akka/persistence/cassandra/query/DirectWriting.scala [hunk 18]
```diff
@@ -26,56 +27,47 @@ trait DirectWriting extends BeforeAndAfterAll {
...
-  private lazy val preparedDeleteMessage = session.prepare(writeStatements.deleteMessage)
+  private lazy val preparedDeleteMessage = cluster.prepare(writeStatements.deleteMessage)
 
   protected def writeTestEvent(persistent: PersistentRepr, partitionNr: Long = 1L): Unit = {
     val event = persistent.payload.asInstanceOf[AnyRef]
     val serializer = serialization.findSerializerFor(event)
     val serialized = ByteBuffer.wrap(serialization.serialize(event).get)
-
+    val nowUuid = Uuids.timeBased()
+    val now = Uuids.unixTimestamp(nowUuid)
```
Comment (helena) [inline core/src/test/scala/akka/persistence/cassandra/query/DirectWriting.scala]: nice

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala [hunk 19]
```diff
@@ -111,18 +97,26 @@ class CassandraSnapshotStore(cfg: Config, cfgPath: String)
...
       else preparedSelectSnapshotMetadata
-    for {
+    val result = for {
       p <- snapshotMetaPs
       mds <- metadata(p, persistenceId, criteria, someMaxLoadAttempts)
       res <- loadNAsync(mds)
     } yield res
+
+    result.onComplete { result =>
+      log.debug("snapshot load complete {}", result)
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala]: Seems weird here, maybe consider just have it in the for comp's `yield`

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala [hunk 20]
```diff
@@ -20,22 +19,14 @@ trait CassandraSnapshotCleanup extends CassandraStatements {
...
-    session
-      .prepare(deleteAllSnapshotForPersistenceId)
-      .map(_.setConsistencyLevel(snapshotConfig.writeConsistency).setIdempotent(true).setRetryPolicy(deleteRetryPolicy))
-
-  def preparedDeleteAllSnapshotsForPidAndSequenceNrBetween =
-    session
-      .prepare(deleteAllSnapshotForPersistenceIdAndSequenceNrBetween)
-      .map(_.setConsistencyLevel(snapshotConfig.writeConsistency).setIdempotent(true).setRetryPolicy(deleteRetryPolicy))
+  def preparedDeleteAllSnapshotsForPidAndSequenceNrBetween: Future[PreparedStatement] =
+    session.prepare(deleteAllSnapshotForPersistenceIdAndSequenceNrBetween)
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: What about the retry policy in these, where is that now from?

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala [hunk 21]
```diff
@@ -428,7 +421,7 @@ import com.github.ghik.silencer.silent
             toSeqNr
         }
         session
-          .selectEventsByPersistenceId(persistenceId, pnr, expectedNextSeqNr, endNr, fetchSize)
+          .selectEventsByPersistenceId(persistenceId, pnr, expectedNextSeqNr, endNr)
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: Where is the fetch size now from?

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala [hunk 22]
```diff
@@ -69,39 +70,33 @@ import com.github.ghik.silencer.silent
       selectEventsByPersistenceIdQuery: PreparedStatement,
       selectSingleRowQuery: PreparedStatement,
       selectDeletedToQuery: PreparedStatement,
-      session: Session,
-      customConsistencyLevel: Option[ConsistencyLevel],
-      customRetryPolicy: Option[RetryPolicy]) {
+      session: CqlSession,
+      profile: String) {
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: I keep wanting some way to validate the profile

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala [hunk 23]
```diff
@@ -31,9 +31,9 @@ import scala.concurrent.ExecutionContextExecutor
  */
 @InternalApi private[akka] final class AllPersistenceIdsStage(
     refreshInterval: Option[FiniteDuration],
-    fetchSize: Int,
```
Comment (helena) [inline core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: Are we just keeping this unconfigured now - because I don't see anything setting `page-size`
https://docs.datastax.com/en/developer/java-driver/4.3/manual/core/paging/#paging

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 21 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: The profile, i tried to remove everything from our config that was in the new driver config

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala (hunk 20 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: The profile :)

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 22 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: as in that we've set it? Or that it exists? We could add something in the CassandraSession to say a profile should be set on a statement

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala (hunk 23 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: :+1:

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 24]
```diff
@@ -641,9 +630,15 @@ class CassandraJournal(cfg: Config, cfgPath: String)
       partitionSize: Long): Future[Long] = {
     def find(currentPnr: Long, currentSnr: Long, foundEmptyPartition: Boolean): Future[Long] = {
       // if every message has been deleted and thus no sequence_nr the driver gives us back 0 for "null" :(
-      val boundSelectHighestSequenceNr = preparedSelectHighestSequenceNr.map(_.bind(persistenceId, currentPnr: JLong))
+      val boundSelectHighestSequenceNr = preparedSelectHighestSequenceNr.map(ps => {
+        log.debug("Binding {} {} to {}", persistenceId, currentPnr)
+        val bound = ps.bind(persistenceId, currentPnr: JLong)
+        log.debug("Binding {} {}  = {}", persistenceId, currentPnr, bound)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: some logging cleanup remaining?

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala (hunk 10 above)
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: should it be `!rs.hasMorePages` to correspond to `rs.isExhausted`?

Comment (patriknw) [reply core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: nice

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 11 above)
Comment (patriknw) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: what settings will be used if profile is not set on the statement? I'm thinking about how it will be for the standalone Alpakka CassandraSession.

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala (hunk 20 above)
Comment (patriknw) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: *then standard answer* ;-)

Comment (patriknw) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: What's the defaults for retries? I think we had some support case where we saw some really long retry timeouts (> 30sec). Anything we need to adjust? It's not good with too long pending writings because then circuit breakers may trigger before the write is completed. Could even result in new actor starting on another node and not seeing the write if it completes after the recovery.

File: session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala [hunk 25]
```diff
@@ -54,133 +50,32 @@ import akka.cassandra.session._
...
+    init: CqlSession => Future[Done])
     extends NoSerializationVerificationNeeded {
-  import settings._
 
   implicit private[akka] val ec = executionContext
   private lazy implicit val materializer = ActorMaterializer()(system)
 
-  // cache of PreparedStatement (PreparedStatement should only be prepared once)
-  private val preparedStatements =
-    new ConcurrentHashMap[String, Future[PreparedStatement]]
```
Comment (patriknw) [inline session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala]: is this built in now?

File: session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala [hunk 26]
```diff
@@ -439,19 +321,20 @@ final class CassandraSession(
         def tryPushOne(): Unit =
           resultSet match {
             case Some(rs) if isAvailable(out) =>
-              if (rs.isExhausted())
-                complete(out)
-              else if (rs.getAvailableWithoutFetching() > 0)
+              if (rs.currentPage().iterator().hasNext) {
```
Comment (patriknw) [inline session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala]: remaining instead?

File: session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala [hunk 27]
```diff
@@ -54,133 +50,32 @@ import akka.cassandra.session._
...
-      } catch {
-        case NonFatal(e) =>
-          // this is only in case the direct calls, such as sessionProvider, throws
-          promise.failure(e)
-      }
-
-    trySetup(settings.connectionRetries)
-
-    promise.future
-  }
```
Comment (patriknw) [inline session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala]: great to get rid of all this

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala (hunk 20 above)
Comment (helena) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: I just have not seen it set in the profile here, hence the clarification.

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 21 above)
Comment (helena) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: So we are using their default then I am guessing.

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 22 above)
Comment (helena) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: I finally clicked into the `.conf` file. At a minimum, perhaps we should consider validating a read/write exist, and a profile value that might be needed for our use that may be different from the default? e.g. a situation like https://github.com/akka/akka-persistence-cassandra/pull/596/files#r354814000. Or a setting that would cause us to fail downstream if not set or set to a size/duration minimum for a vague example.

Say the `datastax-java-driver.basic.request.page-size = 5000`, if too large based on their data, would cause timeout here, would be log it at least now?

File: docs/src/main/paradox/migrations.md (hunk 4 above)
Comment (chbatey) [reply docs/src/main/paradox/migrations.md]: Doh

File: docs/src/main/paradox/migrations.md (hunk 5 above)
Comment (chbatey) [reply docs/src/main/paradox/migrations.md]: Agreed

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 11 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: All the driver's defaults which are defined in its reference.conf, I think that makes sense and not have our own defaults for a generic `CassandraSession`

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala (hunk 19 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala]: Removed, just added when nothing worked with the new driver

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala (hunk 20 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: > I just have not seen it set in the profile here, hence the clarification.

The only thing we override in our profile is the consistency as we want the default to be QUORUM rather than LOCAL_ONE. Everything else is up to the user to override in the profile.

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotCleanup.scala]: > What's the defaults for retries? I think we had some support case where we saw some really long retry timeouts (> 30sec). Anything we need to adjust? It's not good with too long pending writings because then circuit breakers may trigger before the write is completed. Could even result in new actor starting on another node and not seeing the write if it completes after the recovery.

The driver has simplified this and provides no alternative retry policy, but it is still pluggable. 

The default retries once, and only if enough replicas responded but didn't return the data (they start off returning a hash). I think this is a fine default for us and users can plug in a different retry policy if needed

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 21 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: The drivers? Yes

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala (hunk 22 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: > Say the datastax-java-driver.basic.request.page-size = 5000, if too large based on their data, would cause timeout here, would be log it at least now?

I am not sure I understand, how could that ever cause a timeout?

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: > At a minimum, perhaps we should consider validating a read/write exist

As in check it exists in configuration? Sounds senisble

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala]: Added in https://github.com/akka/akka-persistence-cassandra/pull/596/commits/89505be4915c9c4cfe433baba67c0e785aa1578c

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 24 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala]: Heh, i spent quite some time with nothing working adding far too much logging. Removed

File: core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala (hunk 10 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala]: Yep, i switched this to use the common `isExhausted`

File: session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala (hunk 25 above)
Comment (chbatey) [reply session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala]: Yep, the driver cahces prepared statements

File: session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala (hunk 26 above)
Comment (chbatey) [reply session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala]: :+1:

File: session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala (hunk 27 above)
Comment (chbatey) [reply session/src/main/scala/akka/cassandra/session/scaladsl/CassandraSession.scala]: :+1:
